### PR TITLE
Array value configurable. 8,16,32..256 bits (not prefixed to 8)

### PR DESCRIPTION
**Pull request by feliam**
Expression expanded to handle arrays to arbitrary sized values (not just 8 bits)
Previously Array represented smtlib arrays from Bitvec(?) to Bitvec(8). 
With this patch you can now have Arrays of any to any bitsize.  (usable for EVM storage for example)

---
This change is [Reviewable](https://reviewable.io/reviews/trailofbits/manticore/732)

### CODE REVIEW

File: manticore/core/smtlib/constraints.py [hunk 1]
```diff
@@ -163,23 +163,26 @@ def __str__(self):
...
                 raise Exception("Type not supported %r", exp)
             result += '(assert (= %s %s))\n' % (name, smtlib)
 
-        r = translator.pop()
-        while r is not None:
-            result += '(assert %s)\n' % r
-            r = translator.pop()
+        constraint_str = translator.pop()
+        while constraint_str is not None:
+            if not constraint_str is 'true':
```
Comment (disconnect3d) [inline manticore/core/smtlib/constraints.py]: maybe `if constraint_str != 'true':`

Comment (disconnect3d) [reply manticore/core/smtlib/constraints.py]: For sure this shouldn't use `is`, see https://stackoverflow.com/questions/1504717/why-does-comparing-strings-in-python-using-either-or-is-sometimes-produce

File: manticore/core/smtlib/constraints.py [hunk 2]
```diff
@@ -163,23 +163,26 @@ def __str__(self):
...
         return result
 
         buf = ''
         for d in self.declarations:
             buf += d.declaration + '\n'
-        for a in self.constraints:
-            buf += '(assert %s)\n' % translate_to_smtlib(a, use_bindings=True)
+        for constraint in self.constraints:
+            constraint_str = translate_to_smtlib(constraint, use_bindings=True)
+            if not constraint_str is 'true':
```
Comment (disconnect3d) [inline manticore/core/smtlib/constraints.py]: Same as above: `if constraint_str != 'true':` ( https://stackoverflow.com/questions/1504717/why-does-comparing-strings-in-python-using-either-or-is-sometimes-produce )

File: manticore/core/smtlib/expression.py [hunk 3]
```diff
@@ -554,14 +555,23 @@ def cast_value(self, value):
...
         if isinstance(value, (int, long)):
-            return BitVecConstant(8, value)
-        assert isinstance(value, BitVec) and value.size == 8
+            return BitVecConstant(self.value_bits, value)
+        assert isinstance(value, BitVec) and value.size == self.value_bits
         return value
 
+    def __len__(self):
+        if self.index_max is None:
+            raise Exception()
```
Comment (disconnect3d) [inline manticore/core/smtlib/expression.py]: Missing a message :P

File: manticore/core/smtlib/expression.py [hunk 4]
```diff
@@ -616,14 +626,34 @@ def byte(self):
...
+            #fresh array proxy
+            super(ArrayProxy, self).__init__(array.index_bits, array.index_max, array.value_bits)
+            self._array = array
+            self._name = array.name
+        else:
+            #arrayproxy for an prepopulated array
+            super(ArrayProxy, self).__init__(array.index_bits, array.index_max, array.value_bits)
+            self._array = array
+            while not isinstance(array, ArrayVariable):
+                array=array.array
```
Comment (disconnect3d) [inline manticore/core/smtlib/expression.py]: `array = array.array`

**Comment by feliam** [comment]
Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[manticore/core/smtlib/constraints.py, line 173 at r1](https://reviewable.io:443/reviews/trailofbits/manticore/732#-L4MvA1OwogrDNj_eKuL-r1-173:-L5so0nadgyjgonzCsEA:b-896fix) ([raw file](https://github.com/trailofbits/manticore/blob/5ba7634433ffafaea04c1f3e93e2b809b851186a/manticore/core/smtlib/constraints.py#L173)):*
<details><summary><i>Previously, disconnect3d (Disconnect3d) wrote…</i></summary><blockquote>

For sure this shouldn't use `is`, see https://stackoverflow.com/questions/1504717/why-does-comparing-strings-in-python-using-either-or-is-sometimes-produce
</blockquote></details>

Done.

---

*[manticore/core/smtlib/constraints.py, line 184 at r1](https://reviewable.io:443/reviews/trailofbits/manticore/732#-L4MvA1OwogrDNj_eKuL-r1-184:-L5so0JKhUv14vZKWLT6:b-896fix) ([raw file](https://github.com/trailofbits/manticore/blob/5ba7634433ffafaea04c1f3e93e2b809b851186a/manticore/core/smtlib/constraints.py#L184)):*
<details><summary><i>Previously, disconnect3d (Disconnect3d) wrote…</i></summary><blockquote>

Same as above: `if constraint_str != 'true':` ( https://stackoverflow.com/questions/1504717/why-does-comparing-strings-in-python-using-either-or-is-sometimes-produce )
</blockquote></details>

Done.

---

*[manticore/core/smtlib/expression.py, line 564 at r1](https://reviewable.io:443/reviews/trailofbits/manticore/732#-L4MvA1OwogrDNj_eKuM-r1-564:-L5so-k8kE6lWOmwUkm_:b-896fix) ([raw file](https://github.com/trailofbits/manticore/blob/5ba7634433ffafaea04c1f3e93e2b809b851186a/manticore/core/smtlib/expression.py#L564)):*
<details><summary><i>Previously, disconnect3d (Disconnect3d) wrote…</i></summary><blockquote>

Missing a message :P
</blockquote></details>

Done.

---

*[manticore/core/smtlib/expression.py, line 646 at r1](https://reviewable.io:443/reviews/trailofbits/manticore/732#-L4MvA1OwogrDNj_eKuM-r1-646:-L5so-OSTfocdJkHCoz6:b-896fix) ([raw file](https://github.com/trailofbits/manticore/blob/5ba7634433ffafaea04c1f3e93e2b809b851186a/manticore/core/smtlib/expression.py#L646)):*
<details><summary><i>Previously, disconnect3d (Disconnect3d) wrote…</i></summary><blockquote>

`array = array.array`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/trailofbits/manticore/732)*

**Comment by yan** [comment]
Reviewed 5 of 7 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[manticore/core/smtlib/expression.py, line 645 at r2](https://reviewable.io:443/reviews/trailofbits/manticore/732#-L5yCX671jtrtEAG3WKn:-L5yCX671jtrtEAG3WKo:b-euf12i) ([raw file](https://github.com/trailofbits/manticore/blob/cbbadb0991fc4efe9b4b8d1da2771db012e4c2c3/manticore/core/smtlib/expression.py#L645)):*
> ```Python
>             super(ArrayProxy, self).__init__(array.index_bits, array.index_max, array.value_bits)
>             self._array = array
>             while not isinstance(array, ArrayVariable):
> ```

This goes through multiple levels of ArrayProxies, right? Might make sense to add:

```python
def underlying_array(self):
    array = self._array
    while not isinstance(array, ArrayVariable):
        # potentially assert that array is still ArrayProxy
        array = array.array
    return array
```

 to implement this traversal. If this is too heavy-handed, okay without it as well.

---

*[manticore/core/smtlib/expression.py, line 672 at r2](https://reviewable.io:443/reviews/trailofbits/manticore/732#-L5yDnaRsYhsmm6iezXG:-L5yDnaRsYhsmm6iezXH:bkbdv6r) ([raw file](https://github.com/trailofbits/manticore/blob/cbbadb0991fc4efe9b4b8d1da2771db012e4c2c3/manticore/core/smtlib/expression.py#L672)):*
> ```Python
>     @property
>     def value_bits(self):
>         return self._array._value_bits
> ```

Use the property here, not the ivar?

---

*[manticore/core/smtlib/visitors.py, line 73 at r2](https://reviewable.io:443/reviews/trailofbits/manticore/732#-L5yDxBpVqbtwzhPkOsB:-L5yDxBpVqbtwzhPkOsC:b-p2rua3) ([raw file](https://github.com/trailofbits/manticore/blob/cbbadb0991fc4efe9b4b8d1da2771db012e4c2c3/manticore/core/smtlib/visitors.py#L73)):*
> ```Python
>         '''
>         if isinstance(node, ArrayProxy):
>             node = node.array
> ```

Should this only do one indirection or multiple? Potential use for `underlying_array` from above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/trailofbits/manticore/732)*

**Comment by feliam** [comment]
*[manticore/core/smtlib/expression.py, line 645 at r2](https://reviewable.io:443/reviews/trailofbits/manticore/732#-L5yCX671jtrtEAG3WKn:-L62elNfYySrvoHpV6Ej:b-eyicw0) ([raw file](https://github.com/trailofbits/manticore/blob/cbbadb0991fc4efe9b4b8d1da2771db012e4c2c3/manticore/core/smtlib/expression.py#L645)):*
<details><summary><i>Previously, yan (Yan Ivnitskiy) wrote…</i></summary><blockquote>

This goes through multiple levels of ArrayProxies, right? Might make sense to add:

```python
def underlying_array(self):
    array = self._array
    while not isinstance(array, ArrayVariable):
        # potentially assert that array is still ArrayProxy
        array = array.array
    return array
```

 to implement this traversal. If this is too heavy-handed, okay without it as well.
</blockquote></details>

There should be a single ArrayProxy at the outside. Then a bunch of Arraystore and a single ArrayVariable in the deep.
ArrayProxy(ArrayStore(ArrayStore(ArrayVariable(256,256)) ... )))
That @property could make sense for Array. underlying_variable

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/trailofbits/manticore/732)*

**Comment by feliam** [comment]
*[manticore/core/smtlib/visitors.py, line 73 at r2](https://reviewable.io:443/reviews/trailofbits/manticore/732#-L5yDxBpVqbtwzhPkOsB:-L62flaNVLH_fiiBdeEi:b-3dqvw2) ([raw file](https://github.com/trailofbits/manticore/blob/cbbadb0991fc4efe9b4b8d1da2771db012e4c2c3/manticore/core/smtlib/visitors.py#L73)):*
<details><summary><i>Previously, yan (Yan Ivnitskiy) wrote…</i></summary><blockquote>

Should this only do one indirection or multiple? Potential use for `underlying_array` from above.
</blockquote></details>

That underlying array will get you the inner variable not the array. 
Though I'm not 100% confortable with having a special case for ArrayProxy here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/trailofbits/manticore/732)*

**Comment by feliam** [comment]
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[manticore/core/smtlib/constraints.py, line 173 at r1](https://reviewable.io:443/reviews/trailofbits/manticore/732#-L4MvA1OwogrDNj_eKuL-r1-173:-L62czBuporljYV0BTNp:b-896fix) ([raw file](https://github.com/trailofbits/manticore/blob/5ba7634433ffafaea04c1f3e93e2b809b851186a/manticore/core/smtlib/constraints.py#L173)):*
<details><summary><i>Previously, feliam (feliam) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/trailofbits/manticore/732)*

**Comment by feliam** [comment]
*[manticore/core/smtlib/expression.py, line 672 at r2](https://reviewable.io:443/reviews/trailofbits/manticore/732#-L5yDnaRsYhsmm6iezXG:-L62hFO-dt1P_9mXKYNS:b-896fix) ([raw file](https://github.com/trailofbits/manticore/blob/cbbadb0991fc4efe9b4b8d1da2771db012e4c2c3/manticore/core/smtlib/expression.py#L672)):*
<details><summary><i>Previously, yan (Yan Ivnitskiy) wrote…</i></summary><blockquote>

Use the property here, not the ivar?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/trailofbits/manticore/732)*

**Comment by yan** [comment]
:lgtm:

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/trailofbits/manticore/732#-:-L6IrLuuCpkZDMT2peMq:bnfp4nl)*

**Pull request merged**